src/callbacks/save_solution.jl

```diff
@@ -0,0 +1,83 @@
+
+# TODO: Taal, implement, save AMR indicator values
+# TODO: Taal, refactor, allow saving arbitrary functions of the conservative variables
+mutable struct SaveSolutionCallback
+  save_initial_solution::Bool
+  save_final_solution::Bool
+  output_directory::String
+  solution_variables::Symbol
+end
+
+
+function Base.show(io::IO, cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:SaveSolutionCallback}
+  stepsize_callback = cb.affect!
+  print(io, "SaveSolutionCallback")
+end
+# TODO: Taal bikeshedding, implement a method with more information and the signature
+# function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:StepsizeCallback}
+# end
+
+
+function SaveSolutionCallback(; solution_interval=0,
+                                save_initial_solution=true,
+                                save_final_solution=true,
+                                output_directory="out",
+                                solution_variables=:primitive)
+  condition = (u, t, integrator) -> solution_interval > 0 && ((integrator.iter % solution_interval == 0) || (save_final_solution && t == integrator.sol.prob.tspan[2]))
+
+  solution_callback = SaveSolutionCallback(save_initial_solution, save_final_solution,
+                                           output_directory, solution_variables)
+
+  DiscreteCallback(condition, solution_callback,
+                   save_positions=(false,false),
```

Comment:
It tells DifferentialEquations.jl not to save the current solution `(before, after)` the callback is applied.

Reply:
This I don't understand... isn't the callback the one which saves the solution? Or is it saved by DiffEq? Now I am confused.

Reply:
By default, DiffEq creates an array of the solution at intermediate points, similar to MATLAB's `ode45`. That's really nice for small ODEs that you want to visualize interactively. It's a show-stopper for us because of the memory requirements. That's why I used `save_everystep=false` in `solve(ode, ...)`. `save_positions=(false,false)` tells DiffEq to also not save an intermediate solution in RAM before or after the callback is called. By coincidence, this callback will also save a solution, but to a file.

Reply:
:+1: